api/v1/verticadb_webhook.go

```diff
@@ -1701,6 +1760,162 @@ func (v *VerticaDB) checkSandboxPrimary(allErrs field.ErrorList, oldObj *Vertica
+					fmt.Sprintf("cannot unsandbox subcluster %q in sandbox %q that has Shutdown field set to true",
+						oldSubclusterName, oldSandboxName))
+				allErrs = append(allErrs, err)
+				continue
+			}
+		}
+	}
+	return allErrs
+}
+
+// ensure: when a sandbox is to be shutdown, if a subcluster in it has annotation vertica.com/shutdown-driven-by-sandbox set to true, the
+// subcluster's shutdown field should not be updated
+func (v *VerticaDB) validateAnnotatedSubclustersInShutdownSandbox(old runtime.Object, allErrs field.ErrorList) field.ErrorList {
+	oldObj := old.(*VerticaDB)
+	persistSandboxes := v.findPersistSandboxes(oldObj)
+	newSandboxMap := v.GenSandboxMap()
+	oldSandboxMap := oldObj.GenSandboxMap()
+	newSubclusterIndexMap := v.GenSubclusterIndexMap()
+	newSubclusterMap := v.GenSubclusterMap()
+	oldSubclusterMap := oldObj.GenSubclusterMap()
+	for _, sandboxName := range persistSandboxes {
+		newSandbox := newSandboxMap[sandboxName]
+		oldSandbox := oldSandboxMap[sandboxName]
+		if newSandbox.Shutdown {
+			for _, subclusterName := range oldSandbox.Subclusters {
+				oldSubcluster := oldSubclusterMap[subclusterName.Name]
+				if drivenby, ok := oldSubcluster.Annotations["vertica.com/shutdown-driven-by-sandbox"]; ok && drivenby == trueString {
+					newSubcluster, oldSclusterPersist := newSubclusterMap[subclusterName.Name]
+					if oldSclusterPersist && oldSubcluster.Shutdown != newSubcluster.Shutdown {
+						index := newSubclusterIndexMap[subclusterName.Name]
+						p := field.NewPath("spec").Child("subclusters")
+						err := field.Invalid(p.Index(index),
+							subclusterName.Name,
+							fmt.Sprintf("cannot change Shutdown field for subcluster %q that has annotation \"vertica.com/shutdown-driven-by-sandbox\" set to true",
+								subclusterName.Name))
+						allErrs = append(allErrs, err)
+					}
+				}
+			}
+		}
+	}
+	return allErrs
+}
+
+func (v *VerticaDB) validateShutdownSandboxImage(old runtime.Object, allErrs field.ErrorList) field.ErrorList {
+	oldObj := old.(*VerticaDB)
+	persistSandboxes := v.findPersistSandboxes(oldObj)
+	oldSandboxMap := oldObj.GenSandboxMap()
+	newSandboxMap := v.GenSandboxMap()
+	newSandboxIndexMap := v.GenSandboxIndexMap()
+	newSubclusterMap := v.GenSubclusterMap()
+	oldSubclusterIndexMap := oldObj.GenSubclusterIndexMap()
+
+	for _, sandboxName := range persistSandboxes {
+		oldSandbox := oldSandboxMap[sandboxName]
+		newSandbox := newSandboxMap[sandboxName]
+		if oldSandbox.Image != newSandbox.Image {
+			newSandboxIndex := newSandboxIndexMap[sandboxName]
+			errMsgs := make([]string, 0)
+			errMsgs = v.checkSboxForShutdownInSpecAndStatus(newSandbox, newSubclusterMap, oldSubclusterIndexMap, errMsgs)
+			if len(errMsgs) != 0 {
+				p := field.NewPath("spec").Child("sandboxes").Index(newSandboxIndex)
+				err := field.Invalid(p.Child("image"),
+					sandboxName,
+					fmt.Sprintf("cannot change the image for sandbox %q because %q",
+						sandboxName, strings.Join(errMsgs, ",")))
```

Comment:
errMsgs should only get the names of the subclusters that are shut down. Then here we can say "cannot change the image for sandbox %q because its subclusters %v are already shut down. "

Reply:
This piece of code is already refactored and strings.join(0 is removed to address Roy's concern that there are too many err messages.